btf/types.go

```diff
@@ -1142,26 +1148,69 @@ func readAndInflateTypes(r io.Reader, bo binary.ByteOrder, typeLen uint32, rawSt
 
 	for _, dt := range declTags {
 		switch t := dt.Type.(type) {
-		case *Var, *Typedef:
+		case *Var:
 			if dt.Index != -1 {
-				return nil, fmt.Errorf("type %s: index %d is not -1", dt, dt.Index)
+				return nil, fmt.Errorf("type %s: component idx %d is not -1", dt, dt.Index)
 			}
+			t.Tags = append(t.Tags, dt.Value)
+
+		case *Typedef:
+			if dt.Index != -1 {
+				return nil, fmt.Errorf("type %s: component idx %d is not -1", dt, dt.Index)
+			}
+			t.Tags = append(t.Tags, dt.Value)
 
 		case composite:
-			if dt.Index >= len(t.members()) {
-				return nil, fmt.Errorf("type %s: index %d exceeds members of %s", dt, dt.Index, t)
+			if dt.Index >= 0 {
+				members := t.members()
+				if dt.Index >= len(members) {
+					return nil, fmt.Errorf("type %s: component idx %d exceeds members of %s", dt, dt.Index, t)
+				}
+
+				members[dt.Index].Tags = append(members[dt.Index].Tags, dt.Value)
+				continue
+			}
+
+			if dt.Index == -1 {
+				switch t2 := t.(type) {
+				case *Struct:
+					t2.Tags = append(t2.Tags, dt.Value)
+				case *Union:
+					t2.Tags = append(t2.Tags, dt.Value)
+				}
+
+				continue
 			}
 
+			return nil, fmt.Errorf("type %s: decl tag for type %s has invalid component idx", dt, t)
+
 		case *Func:
 			fp, ok := t.Type.(*FuncProto)
 			if !ok {
 				return nil, fmt.Errorf("type %s: %s is not a FuncProto", dt, t.Type)
 			}
 
-			if dt.Index >= len(fp.Params) {
-				return nil, fmt.Errorf("type %s: index %d exceeds params of %s", dt, dt.Index, t)
+			// Ensure the number of argument tag lists equals the number of arguments
+			if len(t.ParamTags) < len(fp.Params) {
+				t.ParamTags = append(t.ParamTags, make([][]string, len(fp.Params)-len(t.ParamTags))...)
 			}
 
+			if dt.Index >= 0 {
+				if dt.Index >= len(fp.Params) {
+					return nil, fmt.Errorf("type %s: component idx %d exceeds params of %s", dt, dt.Index, t)
+				}
+
+				t.ParamTags[dt.Index] = append(t.ParamTags[dt.Index], dt.Value)
+				continue
+			}
+
+			if dt.Index == -1 {
+				t.Tags = append(t.Tags, dt.Value)
+				continue
+			}
+
+			return nil, fmt.Errorf("type %s: decl tag for type %s has invalid component idx", dt, t)
```

Comment:
Minor: can we reject `dt.Index < -1` before entering the type switch? If yes, the code can be simplified
```suggestion
			if dt.Index >= len(fp.Params) {
				return nil, fmt.Errorf("type %s: component idx %d exceeds params of %s", dt, dt.Index, t)
			}

			if len(t.ParamTags) < dt.Index {
				t.ParamTags = make([][]string, len(fp.Params))
			}

			if dt.Index >= 0 {
				t.ParamTags[dt.Index] = append(t.ParamTags[dt.Index], dt.Value)
			} else {
				t.Tags = append(t.Tags, dt.Value)
			}
```

Reply:
I am sorry, but I do not feel like this is an improvement. It does not save me meaningful amount of lines. It makes the error you will get for cases where -1 is the only valid answer less specific ("numbers below -1 are invalid" vs "-1 is the only valid value") and to me having all of the checks together on one screen makes it easier to read than having to scroll to validate all cases are handled.

Reply:
Sure, I see your point. It would still read easier without `continue` (`switch` `case` maybe?) but up to you.

Could you please help me understand why the following code is necessary?
```go
if len(t.ParamTags) < len(fp.Params) {
	t.ParamTags = append(t.ParamTags, make([][]string, len(fp.Params)-len(t.ParamTags))...)
}
```
The number of parameters does not change. We need to allocate a properly-sized array once.